flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -281,10 +289,41 @@ protected void processInput(DefaultActionContext context) throws Exception {
 		if (status == InputStatus.END_OF_INPUT) {
 			context.allActionsCompleted();
 		}
-		else if (status == InputStatus.NOTHING_AVAILABLE) {
+
+		CompletableFuture<?> jointFuture = getInputOutputJointFuture(status);
+		if (jointFuture != null) {
 			SuspendedMailboxDefaultAction suspendedDefaultAction = context.suspendDefaultAction();
-			inputProcessor.getAvailableFuture().thenRun(suspendedDefaultAction::resume);
+			jointFuture.thenRun(suspendedDefaultAction::resume);
```

Comment:
This requires some `ITCase` test - make sure that mailbox thread is not blocked when producing small records.

Reply:
+1